### DISCUSSION TITLE
serial versus parallel sequence of invoke elements in a state

### DISCUSSION THREAD

**Discussion by tfahringer** [alexzhornyak/SCXML-tutorial#28] [Q&A]
Is it correct that a sequence of 
       <invoke> elements
in a specific state is executed by the underlying runtime system in parallel?

so for instance 
<state id="ex">
<invoke type="vxml" ...>
<invoke type="html" ...>
</state>
would make the underlying runtime system to executed them in parallel. Is this correct?

If this is true, then what is an elegant way to specify a sequence of <invoke> elements as part of a specific state in serial, thus they are actually executed serially?
I would like to have a solution that does not put every <invoke> in a separate state and use transition to pass over them serially.

**Comment by alexzhornyak** [comment]
If I understood you correctly, you need to execute submachines in sequence. In this case you need to use auto generated submachine event `done.invoke + SUBMACHINE_ID`

https://user-images.githubusercontent.com/18611095/235148082-93d4881d-7733-45e6-b97b-041cbe364075.mp4

**parent machine**

```xml
<scxml datamodel="ecmascript" name="ScxmlParent" version="1.0" xmlns="http://www.w3.org/2005/07/scxml">
	<state id="child1">
		<invoke autoforward="true" id="ID_1" src="machine_child.scxml"/>
		<transition event="done.invoke.ID_1" target="child2"/>
	</state>
	<state id="child2">
		<invoke autoforward="true" id="ID_2" src="machine_child.scxml"/>
		<transition event="done.invoke.ID_2" target="FinalShape1"/>
	</state>
	<final id="FinalShape1"/>
</scxml>
```

**child machine**
```xml
<scxml datamodel="ecmascript" name="ScxmlChild" version="1.0" xmlns="http://www.w3.org/2005/07/scxml">
	<state id="childStart">
		<onentry>
			<raise event="event"/>
		</onentry>
		<transition event="event" target="childFinish"/>
	</state>
	<final id="childFinish"/>
</scxml>
```

**Comment by tfahringer** [comment]
No, I would like to avoid putting every invoke in a separate state. I would prefer to have all invoke elements in a single atomic state.

question: what is the semantic of 

<scxml datamodel="ecmascript" name="ScxmlParent" version="1.0" xmlns="http://www.w3.org/2005/07/scxml">
	<state id="child1">
		<invoke type ="vxml" id="voice" src="..."  />
		<invoke type ="HTML" id="GUI" src=""  />
                <invoke type =...  />
	</state>

Are these 3 invoke elements executed in parallel or in serial based on the standard?

I would like to avoid any solution that puts these 3 invokes in separate states!!!

**Reply by alexzhornyak** [reply]
Invoked element is a separate session, it even could be loaded at another network place. It is working in parallel with its own event queue until the command from the parent state machine to stop execution. If you do not want to have invoked processes but you need the same logic in single chart you have to make unique names.

Take a [look at this comparison example](https://github.com/alexzhornyak/SCXML-tutorial/blob/master/Doc/invoke.md#1-unoptimized-statechart-without-invoke)

**Comment by tfahringer** [comment]
None of these solutions are good.  As I said I do NOT a solution that puts every invoke element in a separate state. Your example above puts all invoke element in separate states which I dont want.

There is the paper "Introduction to SCXML" by Jim Barnett.  On page 93 it shows an example with 2 invokes in a single state.

<state id=”address”>
<invoke type=”vxml” id=”voice” src=”http://somestore.com/orders/getAddress.vxml/”>
<invoke type=”html” id=”GUI” src=”http://somestore.com/orders/getAddress.html/”>
<transition event=”done.invoke.voice” target=”continueVoiceInput”/>
<transition event=”done.invoke.GUI” target=”continueGUIInput”
<transition event=”cancel” target=”terminateOrder”/>
</state>

What is the semantic of this? Are there 2 separate sessions started - one for every invoke?
My main question for this example is whether these 2 sessions are started and executed in parallel? Or is the first invoke first finished before the next one is started?

**Reply by alexzhornyak** [reply]
![image](https://user-images.githubusercontent.com/18611095/235183998-f695d952-6da1-4063-932b-4757ae584aad.png)

![image](https://user-images.githubusercontent.com/18611095/235184184-7de7beff-4f03-463b-8380-44717317fcc1.png)

## Explanation:
It we are talking about invoked **SCXML Event/IO processor** for example in USCXML framework every invoked state machine is the separate thread.

So after enter state **address**  state machine **Voice** will be invoked and after the time when it reaches the stable state, state machine "GUI" will be invoked.

After the reaching stable state there will be 3 threads:
- thread main
- thread Voice
- thread GUI

They will process events in their own event queues.

After the time when state machine Voice will reach the final state it will generate **done.invoke.voice** event and thread Voice will be closed, Same to thread GUI

**Reply by alexzhornyak** [reply]
Also once we will exit state **address** for example by event **cancel** threads Voice and GUI will be terminated

**Comment by tfahringer** [comment]
[Invoke1.txt](https://github.com/alexzhornyak/SCXML-tutorial/files/11354715/Invoke1.txt)

**Comment by tfahringer** [comment]
what actually means stable state.

I read the language specification but cannot really figure out what that means.

Let's assume that SM voice does a lot of processing for 1 hour or so by invoking external web services or whatever. does this mean that SM GUI has to wait for that until it will be invoked?

It is important for me that I can invoke some compute task inside a state. This can be a web service, serverless functions or whatever. most likely a program that runs outside of SCXML. But it does a lot of data processing.

Then the question is when is this state with this processing task stable? Before it starts this task or afterwards?

**Reply by alexzhornyak** [reply]
**Stable state** means that event queue of the state machine is empty and it is waiting for incoming events.

The case that I told you about USCXML is only because of local PC architecture when thread-blocking task will block other threads for execution

![image](https://user-images.githubusercontent.com/18611095/235192266-c07b8d31-6691-4818-a1f5-91e0914f9cea.png)

If you invoked service will be an external Web, cloud etc. , it will be executed in parallel and will not wait for stable states etc.

The only thing that I will again tell that once you will exit parent state you should terminate child invoked sessions
